File: CC OffSite/TME.CarConfigurator.Administration/Extensions/ColouringModesExtensions.vb
Imports System.Runtime.CompilerServices
Imports System.Collections.Generic
Imports TME.CarConfigurator.Administration.Enums

Namespace Extensions
    Public Module ColouringModesExtensions
        <Extension>
        Public Function AsString(ByVal colouringModes As ColouringModes) As String
            If colouringModes = colouringModes.None Then Return "None"
            If colouringModes = colouringModes.BodyColour Then Return "Body Colour"
            If colouringModes = colouringModes.PrimaryAccentColour Then Return "Primary Accent Colour"
            If colouringModes = colouringModes.SecondaryAccentColour Then Return "Secondary Accent Colour"

            Dim list = New List(Of String)()
            If (colouringModes And colouringModes.BodyColour) = colouringModes.BodyColour Then list.Add("Body")
            If (colouringModes And colouringModes.PrimaryAccentColour) = colouringModes.PrimaryAccentColour Then list.Add("Primary Accent")
            If (colouringModes And colouringModes.SecondaryAccentColour) = colouringModes.SecondaryAccentColour Then list.Add("Secondary Accent")

            Dim result = String.Join(", ", list)
            Dim lastSeperator = result.LastIndexOf(", ", StringComparison.Ordinal)
            result = result.Remove(lastSeperator, 2).Insert(lastSeperator, " or ")
            result = result + " Colour"
            Return result

        End Function
    End Module
End Namespace

File: CC OffSite/TME.CarConfigurator.Publisher/Extensions/ColouringModesExtensions.cs
using TME.CarConfigurator.Administration.Enums;
using TME.CarConfigurator.Publisher.Exceptions;

namespace TME.CarConfigurator.Publisher.Extensions
{
    public static class ColouringModesExtensions
    {
        public static Repository.Objects.Enums.ColouringModes Convert(this ColouringModes modes)
        {
            if ((modes & ~ColouringModes.SecondaryAccentColour & ~ColouringModes.PrimaryAccentColour & ~ColouringModes.BodyColour & ~ColouringModes.None) != 0)
                throw new UnrecognisedColouringModeException(modes);

            Repository.Objects.Enums.ColouringModes result = 0;
            if (modes.HasFlag(ColouringModes.SecondaryAccentColour))
                result |= Repository.Objects.Enums.ColouringModes.SecondaryAccentColour;
            if (modes.HasFlag(ColouringModes.PrimaryAccentColour))
                result |= Repository.Objects.Enums.ColouringModes.PrimaryAccentColour;
            if (modes.HasFlag(ColouringModes.BodyColour))
                result |= Repository.Objects.Enums.ColouringModes.BodyColour;

            return result;
        }
    }
}
